### DISCUSSION TITLE
Help getting cloud signing working with notarization

### DISCUSSION THREAD

**Discussion by OwenLaippleSWF** [fastlane/fastlane#22234] [Q&A]
I'm trying to build, sign and notarize my desktop macos application that will be distributed outside of the appstore.

```
default_platform(:mac)

platform :mac do
  desc "Build and notarize macOS app"
  lane :build do
    unlock_keychain(
      path: "/Users/username/Library/Keychains/login.keychain-db",
      password: "mypassword"
    )

    api_key = app_store_connect_api_key(
      key_id: "my_key_id",
      issuer_id: "my_issuer_id",
      key_filepath: "path_to_my_key",
      duration: 1200,
      in_house: false
    )

    absolute_key_path = File.expand_path("path_to_my_key")

    gym(
      scheme: "myscheme",
      project: "myproject.xcodeproj",
      configuration: "Release",
      clean: true,
      output_name: "myproject",
      skip_package_ipa: true,
      skip_package_pkg: true,
      export_method: "developer-id",
      xcargs: "-allowProvisioningUpdates -authenticationKeyIssuerID #{api_key[:issuer_id]} -authenticationKeyID #{api_key[:key_id]} -authenticationKeyPath #{absolute_key_path}"
    )

    notarize(
      package: "./myproject",
      bundle_id: "com.my.project.foobar",
      print_log: true,
      api_key: api_key
    )
  end
end

```

I get      "message": "The binary is not signed with a valid Developer ID certificate.",
and "      "message": "The signature does not include a secure timestamp."," this should all be handled by cloud signing but I couldn't find much documentation on using it besides the xcargs. When building, signing, and notarizing inside of xcode on the same machine it works fine.

**Comment by bella964** [comment]
Hello,
The errors you're encountering indicate that the app is either not being signed with the correct Developer ID certificate or lacks a secure timestamp. Here's how to address these issues:

Steps to Resolve:
Validate Your Certificate:

Open Keychain Access and ensure a valid Developer ID Application certificate is installed in the login keychain and marked as trusted.
Verify the certificate is not expired or revoked.
Secure Timestamp:

A secure timestamp is added during the signing process to validate the app's authenticity over time. Ensure your signing process includes this step.
Ensure Correct Provisioning Profiles:

Confirm that the bundle ID of your app matches the provisioning profile associated with the Developer ID certificate.
Check Keychain Unlocking:

Ensure the keychain is properly unlocked during the process and that the signing certificate is accessible.
Validate Build Settings:

Double-check that your build configuration (e.g., "Release") is set to use manual signing with the correct Developer ID certificate.
Inspect the Signed App:

After signing, use the following command to verify the app:
bash
Copy code
codesign -dvvv --deep --strict ./myproject
Look for:
The correct Developer ID Application certificate.
A secure timestamp.
Notarization Input:
[JPMorgan Chase Bank Card](https://www.jpmcbcard.com)
Ensure you are submitting the correct package for notarization:
Zip the .app before notarizing.
Use the correct bundle ID.
Test the Process in Xcode:

If the app builds, signs, and notarizes successfully in Xcode, replicate those steps in your custom setup.
Enable Debugging:

Use verbose logs during the build, signing, and notarization processes to identify specific issues.
Manually Test Signing and Notarization:

Sign the app manually with the codesign command and notarize with xcrun altool to isolate Fastlane-related issues.
Common Causes:
Missing or incorrect Developer ID Application certificate.
Misconfigured signing options or provisioning profiles.
Not including all binaries, frameworks, or resources in the signing process.
Submitting an incorrect file (e.g., unzipped .app) for notarization.
